golem/managers/work/concurrent.py

```diff
@@ -1,31 +1,46 @@
 import asyncio
 import logging
-from typing import List
+from typing import Awaitable, Callable, List
 
-from golem.managers.base import DoWorkCallable, Work, WorkManager, WorkResult
+from golem.managers.base import Work, WorkContext, WorkManager, WorkResult
 from golem.managers.work.mixins import WorkManagerPluginsMixin
 from golem.node import GolemNode
+from golem.resources import Activity
 from golem.utils.asyncio import create_task_with_logging
 from golem.utils.logging import trace_span
 
 logger = logging.getLogger(__name__)
 
 
 class ConcurrentWorkManager(WorkManagerPluginsMixin, WorkManager):
-    def __init__(self, golem: GolemNode, do_work: DoWorkCallable, size: int, *args, **kwargs):
-        self._do_work = do_work
+    def __init__(
+        self,
+        golem: GolemNode,
+        get_activity: Callable[[], Awaitable[Activity]],
+        size: int,
+        *args,
+        **kwargs,
+    ):
+        self._get_activity = get_activity
         self._size = size
 
         self._queue: asyncio.Queue[Work] = asyncio.Queue()
         self._results: List[WorkResult] = []
 
         super().__init__(*args, **kwargs)
 
-    @trace_span(show_arguments=True, show_results=True)
     async def do_work(self, work: Work) -> WorkResult:
-        result = await self._do_work_with_plugins(self._do_work, work)
-        logger.info(f"Work `{work}` completed")
-        return result
+        activity = await self._get_activity()
+        work_context = WorkContext(activity)
+        try:
+            work_result = await work(work_context)
+        except Exception as e:
+            work_result = WorkResult(exception=e)
+        else:
+            if not isinstance(work_result, WorkResult):
+                work_result = WorkResult(result=work_result)
+        await activity.destroy()
+        return work_result
```

Comment:
Also this portion of code looks exactly the same in both our work managers cases - maybe its good to move it to parent as a default implementation? Plugin mixin can override this method to wrap work with plugins and run default implementation.